cmd/basic/main.go

```diff
@@ -78,152 +82,204 @@ func main() {
+	ZipfV          float64 // v (< keySpace) puts the main part of the curve before this number
+	tachymeter     *tachymeter.Tachymeter
 }
 
-func (l *BasicLoader) Run() {
-	var runners int
-	// TODO: should be method of surfacing errors.
-	doneChan := make(chan int, 50)
+func (conf *Config) Run() (err error) {
+	g, _ := errgroup.WithContext(context.Background())
 
-	for {
-		for runners < l.desiredConnCount {
-			go l.Worker(doneChan)
-			runners++
-		}
-		res := <-doneChan
-		if res == 0 {
-			//fmt.Println("That's a bingo!")
-		}
-		runners--
-		if *cpuprofile != "" && time.Now().After(l.stopAfter) {
+	samples := conf.RPS * conf.ConnCount
+	if samples < 1000 {
+		samples = 1000
+	}
+
+	if conf.WarmPercent > 0 {
+		err = conf.WarmCache()
+		if err != nil {
 			return
 		}
 	}
+
+	threadStats := make(chan Stats, conf.ConnCount)
+	conf.tachymeter = tachymeter.New(&tachymeter.Config{Size: samples})
+	startTime := time.Now()
+
+	for worker := 0; worker < conf.ConnCount; worker++ {
+		index := worker
+		g.Go(func() error {
+			return conf.Worker(index, threadStats)
+		})
+	}
+
+	err = g.Wait()
+	endTime := time.Now()
+	if err != nil {
+		return
+	}
+
+	conf.tachymeter.SetWallTime(time.Since(startTime))
+	close(threadStats)
+	testStats := &Stats{}
+	for stats := range threadStats {
+		testStats.Add(&stats)
+	}
+
+	report := &Report{
+		StartTime: startTime,
+		EndTime:   endTime,
+		Config:    conf,
+		Metrics:   conf.tachymeter.Calc(),
+		Stats:     testStats,
+	}
+	err = report.PrettyPrint()
+
+	return
 }
 
-func (l *BasicLoader) Timer(tag string, start time.Time) {
-	duration := time.Since(start)
-	if duration > time.Millisecond * 10 {
-		fmt.Printf("%s [%d]\n", tag, int64(time.Since(start) / time.Microsecond))
+func (conf *Config) WarmCache() error {
+	mc := mct.NewClient(conf.Servers[0], conf.Socket, conf.Pipelines, conf.KeyPrefix, conf.StripKeyPrefix)
+
+	rs := pcgr.New(conf.RngSeed, 0)
+	randR := rand.New(&rs)
+	subRS := pcgr.New(1, 0)
+
+	for keyIndex := 0; keyIndex < conf.KeySpace; keyIndex++ {
+		if randR.Intn(100) < conf.WarmPercent {
+			subRS.Seed(conf.RngSeed + int64(keyIndex))
+			key := mct.RandString(&subRS, conf.KeyLength, conf.KeyPrefix)
+
+			valSeed := new(big.Int).SetBytes([]byte(key)).Int64()
+			subRS.Seed(valSeed)
+			value := mct.RandBytes(&subRS, int(conf.ValueSize))
+
+			_, err := mc.Set(key, uint32(conf.ClientFlags), uint32(conf.KeyTTL), value)
+			if err != nil {
+				fmt.Println(err)
+				return err
+			}
+		}
 	}
+
+	return nil
 }
 
-// TODO: use sync.Pool for Item/etc?
-// pool.Put() items back before sleep.
-// may also be able to cache mc's bufio's this way.
-func (l *BasicLoader) Worker(doneChan chan<- int) {
-	// FIXME: selector.
-	host := l.servers[0]
-	mc := mct.NewClient(host)
-	bundles := l.requestBundlesPerConn
+func (conf *Config) Worker(index int, results chan Stats) error {
+	workerSeed := conf.RngSeed + int64(index) + int64(conf.KeySpace)
```

Comment:
Each goroutine is given an index, which is used to compute a unique seed for that worker. This ensures that the sequences of operations is unique for each worker, but still deterministic based on the global seed.